### ISSUE TITLE
Wrong version of newtonsoft??

### ISSUE THREAD

**Issue by JamesC222** [XeroAPI/Xero-NetStandard#189]
Version 3.0.0

The nuget package documentation says it depends on newtonsoft 12 but my programmers tell me its code refers to newtonsoft 11. 

![image](https://user-images.githubusercontent.com/8457735/84270791-23bcc400-ab23-11ea-877c-a6d252c8df7e.png)

**Comment by jenksguo** [comment]
Thanks @jamescasserly, we will look into this and correct the version number if indeed is inconsistent.

**Comment by jamescasserly** [comment]
Apparently your Oauth2client refers to IdentityModel 3.0 and IdentityModel refers to newtonsoft 11.

**Comment by jenksguo** [comment]
@JamesC222 looks like the IdentityModel's [latest stable version 4.3.1](https://www.nuget.org/packages/IdentityModel/4.3.1) does not even use Newtonsoft 12. 

The [latest beta version 5.0.0-preview](https://www.nuget.org/packages/IdentityModel/5.0.0-preview.0) uses System.Text.Json. I can look into adopting it once it is out of beta. 

Would you be able to make Newtonsoft 12 work with the latest [OAuth2Client v1.1.0](https://www.nuget.org/packages/Xero.NetStandard.OAuth2Client/1.1.0)?

**Comment by jkallman** [comment]
I ran into this dependency issue when deploying a .net core 3.1.1 application to IIS 10. 

**tl;dr - Disable `ReadyToRun` when deploying to IIS** `<PublishReadyToRun>False</PublishReadyToRun>`

--

When building the application locally, it would include Newtonsoft.Json 12.0.1 and everything ran perfectly. However, once deploying the application to IIS, it would include Newtonsoft.Json 11.0.2. If you disable `ReadtoRun`, then it will include the correct dependency.

I like your idea @jenksguo of adopting System.Text.Json and upgrading to IdentityModel v5 when released. I don't think there is anything else that can be done until then, but I wanted to provide a workaround for anyone else who may be in the same boat.

**Exception**
I'm including the exception thrown to help anyone googling the same scenario:

    System.IO.FileNotFoundException: Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'. The system cannot find the file specified.
      Module "Xero.NetStandard.OAuth2.Client.CustomJsonCodec", in .ctor
      Module "Xero.NetStandard.OAuth2.Client.ApiClient", in newRequest
      Module "Xero.NetStandard.OAuth2.Client.ApiClient+<GetAsync>d__9`1", in MoveNext
      Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
      Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
      Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
      Module "System.Runtime.CompilerServices.TaskAwaiter`1", in GetResult
      Module "Xero.NetStandard.OAuth2.Api.AccountingApi", in GetOrganisationsAsyncWithHttpInfo
      Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
      Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
      Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
      Module "System.Runtime.CompilerServices.TaskAwaiter`1", in GetResult
      Module "Xero.NetStandard.OAuth2.Api.AccountingApi", in GetOrganisationsAsync
      Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
      Module "System.Runtime.CompilerServices.TaskAwaiter", in ThrowForNonSuccess
      Module "System.Runtime.CompilerServices.TaskAwaiter", in HandleNonSuccessAndDebuggerNotification
      Module "System.Runtime.CompilerServices.TaskAwaiter`1", in GetResult

**Comment by jenksguo** [comment]
@steve34581133 thanks for your feedback on this. I have take another look at all of your responses. 

I will look into two solutions very soon: 
1. downgrade Xero.NetStandard.OAuth2's NewtonSoft dependency to 11
2. move Xero.NetStandard.OAuth2 package to  System.Text.Json

**Comment by jenksguo** [comment]
Hey @steve34581133 @JamesC222 @jkallman, I am considering downgrading NewtonSoft to 10.0.2 on the main API SDK. Would you be able to test if this way your environments like it?

**Comment by steve34581133** [comment]
Hi Jenksguo,

I'm not sure that the downgrade to Newtonsoft 10.0.2 will work in my Azure Automation Account environment. 

If I understand correctly, downgrading Newtonsoft.Json to 10.0.2 for Xero.Netstandard.OAuth2 will leave in place a dependency on Newtonsoft.Json 11.0.2 in Xero.NetStandard.OAuth2Client. I'm willing to try but I'm pretty sure that because the two Xero SDK dlls need different versions of the same Newtonsoft.Json.dll, my PowerShell script will fail when execution hits a line of code that needs the version of the dll that is not present, much the same problem as before.

Would it be possible to downgrade the main API SDK to Newtonsoft.Json 11.0.2 so that it matches the version used by IdentityModel? 


Note for clarification: Xero.NetStandard.OAuth2Client currently uses IdentityModel >= 4.0.0 which has a dependency on Newtonsoft.Json >= 11.0.2. Xero.NetStandard.OAuth2 currently has a dependency on Newtonsoft.Json >= 12.0.1. When Xero API SDK .Net dll methods are called in an Azure PowerShell environment a search is made for the dependency dlls. Unlike the GAC only one version of a dll file can exist in Azure's Automation Account environment, so when a method is called on the dll that was not installed a FileNotFoundException is thrown. Unfortunately it doesn't seem possible to control binding redirection of .Net dlls from within an Azure PowerShell environment.

**Comment by jenksguo** [comment]
@steve34581133 I had a typo in my earlier message, sorry. 

I meant to suggest downgrading the main SDK client Xero.NetStandardOAuth2's Newtonsoft to 11.0.2 to align with identity model 4.0.0. 

When identity model is upgraded to 5.0.0 which uses System.Text.Json, I think I can still keep using Newtonsoft for Xero.NetStandardOAuth2 because there will be nolonger conflicts in dll. Do you agree with this?

**Comment by JamesC222** [comment]
Once we figured this out we solved it by making a binding redirection in our web.config.
            <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral" />
            <bindingRedirect oldVersion="0.0.0.0-12.0.0.0" newVersion="12.0.0.0" />
Not pretty, but I can not say much more about the best way for you to go forward with this.

**Comment by steve34581133** [comment]
Hi Jenksguo,

Yes, this sounds good to me. Once the two Newtonsoft dependencies are matched my problem should be resolved.

**Comment by jenksguo** [comment]
@JamesC222 @steve34581133 Thanks for your inputs. 

I don't see a huge issues in downgrading the Newtonsoft versions and there is no breaking changes either. 

Initial tests suggests it is fine. once full tests are done. I will make a release.

**Comment by jenksguo** [comment]
This should be addressed in v3.6.0 now, @steve34581133 can you please test in your environment and let me know?

**Comment by steve34581133** [comment]
@jenksguo After limited testing I believe my problem with NewtonSoft versions to be resolved. I'm now hitting some kind of new authentication issue or perhaps a problem with how I have constructed my PowerShell script, so I'm yet to make a successful API call. But this is the furthest I have managed to get, and I'm pretty sure my new problems are not related to NewtonSoft.
Thanks for your help in getting this far.

**Comment by jenksguo** [comment]
Thanks for testing @steve34581133. I will close this now. If anything else please raise a new issue.
